overlord/servicestate/servicestate.go

```diff
@@ -42,14 +44,206 @@ import (
+	case "all":
+		us.Selector = UserSelectionAll
+	default:
+		return fmt.Errorf(`cannot unmarshal, expected one of: "self", "all"`)
+	}
+	return nil
+}
+
+type ScopeSelector []string
+
+func (ss *ScopeSelector) UnmarshalJSON(b []byte) error {
+	var scopes []string
+	if err := json.Unmarshal(b, &scopes); err != nil {
+		return fmt.Errorf("cannot unmarshal, expected a list of strings")
+	}
+
+	if len(scopes) > 2 {
+		return fmt.Errorf("unexpected number of scopes: %v", scopes)
+	}
+
+	for _, s := range scopes {
+		switch s {
+		case "system", "user":
+		default:
+			return fmt.Errorf(`cannot unmarshal, expected one of: "system", "user"`)
+		}
+	}
+	*ss = scopes
+	return nil
+}
+
 type Instruction struct {
-	Action string   `json:"action"`
-	Names  []string `json:"names"`
+	Action string        `json:"action"`
+	Names  []string      `json:"names"`
+	Scope  ScopeSelector `json:"scope"`
+	Users  UserSelector  `json:"users"`
 	client.StartOptions
 	client.StopOptions
 	client.RestartOptions
 }
 
+func (i *Instruction) ServiceScope() wrappers.ServiceScope {
+	var hasUser, hasSystem bool
+	for _, opt := range i.Scope {
+		switch opt {
+		case "user":
+			hasUser = true
+		case "system":
+			hasSystem = true
+		}
+	}
+	switch {
+	case hasSystem && !hasUser:
+		return wrappers.ServiceScopeSystem
+	case hasUser && !hasSystem:
+		return wrappers.ServiceScopeUser
+	default:
+		return wrappers.ServiceScopeAll
+	}
+}
+
+func (i *Instruction) hasUserService(apps []*snap.AppInfo) bool {
+	for _, app := range apps {
+		if app.IsService() && app.DaemonScope == snap.UserDaemon {
+			return true
+		}
+	}
+	return false
+}
+
+// EnsureDefaultScopeForUser sets up default scopes based on the type of user
+// if none were provided.
+// Make sure to call Instruction.Validate() before calling this.
+func (i *Instruction) EnsureDefaultScopeForUser(u *user.User) {
+	// Set default scopes if not provided
+	if len(i.Scope) == 0 {
+		// If root is making this request, implied scopes are all
+		if u.Uid == "0" {
+			i.Scope = ScopeSelector{"system", "user"}
+		} else {
+			// Otherwise imply the service scope only
+			i.Scope = ScopeSelector{"system"}
+		}
+	}
+}
+
+func (i *Instruction) validateScope(u *user.User, apps []*snap.AppInfo) error {
+	if len(i.Scope) == 0 {
+		// Providing no scope is only an issue for non-root users if the
+		// target is user-daemons.
+		if u.Uid != "0" && i.hasUserService(apps) {
+			return fmt.Errorf("non-root users must specify service scope when targeting user services")
+		}
+	}
+	return nil
+}
+
+func (i *Instruction) validateUsers(u *user.User, apps []*snap.AppInfo) error {
+	users, err := i.Users.UserList(u)
+	if err != nil {
+		return err
+	}
+
+	// Perform some additional user checks
+	if len(users) == 0 {
+		// It is an error for a non-root to not specify any users if we are targeting
+		// user daemons
+		if u.Uid != "0" && i.hasUserService(apps) {
+			return fmt.Errorf("non-root users must specify users when targeting user services")
```

Comment:
won't this block  the user using users="all" which results in users being nil?

Reply:
Super nice catch! I added a missing unit test and also fixed the check.